src/awkward/contents/listoffsetarray.py

```diff
@@ -153,6 +165,10 @@ def _repr(self, indent, pre, post):
         return "".join(out)
 
     def to_ListOffsetArray64(self, start_at_zero=False):
+        if not self._backend.nplike.known_data:
```

Comment:
Just passing along the `offsets` doesn't count as "touching" them. This kind of subtlety is what motivated this low-level implementation, so that we can make free choices at this level that give us the desired effects.

`ListOffsetArray.to_ListOffsetArray64` makes some value-dependent choices, and the decision to touch or not touch can't depend on values. (Note: the return type does not depend on values—that much is okay.)

I guess there's one case that we can determine from types and metadata only in which the `offsets` could go untouched:

  * if the `ListOffsetArray` already has 64-bit `offsets`
  * if `start_at_zero=False`

The case in which `start_at_zero=True` and `offsets[0] == 0` can't go untouched because we can't know that with a type-tracer.

I'm not going to try threading the touch-logic through the implementation, but I'll put this special case in the touch-logic block on lines 168‒170. It's in 98d0cce5cdc7740912d67ce94e872e55aee75ab2.

Reply:
The bottom line is whether Dask needs to go fetch data to populate the buffer at execution-time. Users (Coffea in particular) want that to happen as little as possible. To put your example in code,

```python
original_data = some_data_source.download_stuff("htts://source.cern")

a_new_array = original_data["list_offset_array_3"]

array_with_parameter = ak.with_parameter(a_new_array, "some_parameter", "some_value")

actually_compute_something = function(original_data["another_field"])

result = ak.zip({"one": array_with_parameter, "two": actually_compute_something})
```

If this is the entire Dask graph (`result.compute()`), my thinking was that we don't _want_ to download the data with field `"list_offset_array_3"` because we never do anything with it. But you bring up a good point that if `"list_offset_array_3"` is removed from the Dask graph, this code will fail in the executor. I hadn't considered that.

One way out is to say that the set of touched input buffers are not removed from the Dask graph: the Dask graph is untouched (_not_ what we were planning!) and instead, the unused buffers are filled with dummy data. NumPy can make very lightweight dummy data:

```python
>>> dummy = np.broadcast_to(np.float64(0), 100)
>>> dummy
array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])
>>> dummy.shape, dummy.strides
((100,), (0,))
```

The buffer associated with the above only has 8 bytes, not 800.

_On the other hand,_ maybe if somebody makes a dask-awkward array like `result`

```python
result = ak.zip({"one": array_with_parameter, "two": actually_compute_something})
```

and then ask for

```python
materialized = result.compute()
```

then they want `materialized["one"]` to contain downloaded data, not a dummy placeholder. They wrote the above code, deliberately, to get `"list_offset_array_3"` as `"one"` with a parameter. In this interpretation (which I'm thinking is the right one, now that I'm going through it), `"list_offset_array_3"` _should_ be included in the Dask graph and _should_ be downloaded.

The way that dask-awkward can get that is by doing this in the last step of report-collecting:

```python
result.meta.layout._touch_data(recursive=True)
```

where `result.meta.layout` is the `Content` containing `TypeTracerArrays`. Everything that makes its way into the final `result` should be touched because the user is explicitly asking for it. If there's another `"list_offset_array_2"` field in the `original_data`, that would be excluded because it doesn't get to the final `result`.

If we do things that way, `"list_offset_array_3"` does not get touched in the `ak.with_parameter` call, but does get touched because it's in the output. If there was another line in the code,

```python
unused = ak.with_parameter(original_data["list_offset_array_2"], "x", "y")
```

this `unused` wouldn't be part of the Dask graph if it's not added to the `result` that gets computed, and so it's important that `ak.with_parameter` does not, by itself, touch data.

Coffea users would have to be careful:

```python
nanoaod = uproot.dask("root://source.cern")
nanoevents = ak.zip({
    "muons": ak.zip({
        "pt": nanoaod["Muon_pt"],
        "phi": nanoaod["Muon_phi"],
        "eta": nanoaod["Muon_eta"],
        "charge": nanoaod["Muon_charge"],
    }),
})

materialized = nanoevents.muons.compute()
```

would cause all muon fields to be downloaded while

```python
materialized = nanoevents.muons[nanoevents.pt > 10].eta
```

would only cause `"Muon_pt"` and `"Muon_eta"` to be downloaded. _Not_ being specific about what you want causes _more_ to be downloaded.

Cc: @douglasdavis, @martindurant, @lgray, and @nsmith-

Reply:
The above doesn't imply any changes in how this PR is implemented, just how it is used. I think the PR is mergeable in its current state. Do you agree, @agoose77?

Reply:
Well, @agoose77 brought up another few cases that make it clear we'll need to do both of the things described above: we'll need dummy buffers _and_ dask-awkward will need to recursively touch the final `result` (the array that gets `.compute()`d).

**Here's case 1:**

```python
original = go_get_data()

zipped = ak.zip({"x": original["field1"], "y": original["field2"], "z": original["field3"]})

result = zipped.x + zipped.y

materialized = result.compute()
```

When dask-awkward gathers a report, it would find that only `"field1"` and `"field2"` have been touched, and that's good because these are the only buffers needed to produce the final `result`. However, if `"field3"` is excluded from the Dask graph, executors would fail to compute the `ak.zip`. The `"field3"` data is not needed, but the metadata is.

**Here's case 2:**

```python
original = go_get_data()   # type: list(X)

result = ak.num(original)

materialized = result.compute()
```

When dask-awkward gathers a report, it would find that only the `offsets` buffer has been touched, not the `content` of the top-level list node, and that's good because the `content` can include anything, such as a large record array, and none of that is in the final `result`. However, Awkward Array can't build a list array node without some kind of `content`.

Both of the above problems can be solved by using the set of touched buffers to determine which buffers need to be downloaded/read from the source, with the others filled in with a dummy. As I said above, NumPy is capable of making very lightweight dummies with arbitrary `shape` (to satisfy metadata constraints). Using the report this way, however, is entirely different from the dask-awkward implementation that we've been pursuing so far: this leaves the Dask graph untouched (i.e. no `project_columns`).

This needs to be done _in addition_ to the recursive-touch described above, because **there's still case 3:**

```python
original = go_get_data()

result = ak.zip({"xy": original["field1"] + original["field2"], "z": original["field3"]})

materialized = result.compute()
```

Unless dask-awkward explicitly calls `result.meta.layout._touch_data(recursive=True)`, the `materialized` result would not include `"field3"`, and it needs to include `"field3"` because a user might subsequently access

```python
materialized.z
```

and expect to not get a dummy buffer.

So unfortunately, this approach would be a big left-turn: different from what you've been developing so far. However, I think it solves all of the problems.

Note that dummy buffers aren't going to cause any performance issues, calculating lots of zeros for no reason. **For example, in case 4:**

```python
original = go_get_data()

zipped = ak.zip({"x": original["field1"], "y": original["field2"], "z": original["field3"]})

unused = ackermann(zipped.z)

result = zipped.x + zipped.y

materialized = result.compute()
```

The `"field3"` buffer would be a dummy in the Dask executor, but the Ackermann function would _not_ be executed on it because it's not part of the Dask graph represented by `result`. When you do the report-generation, you're walking over the operations defined in the Dask graph only.

------------------

This will likely affect all users of dask-awkward, such as `uproot.dask` (cc @kkothari2001), since they'll need to do something different from implementing `project_columns` to reduce the number of array-reads. They'll need to reduce the number of array-reads and replace them with dummy buffers. Somebody (Awkward? dask-awkward?) should provide a standard function for this so that it doesn't get reinvented multiple times.

Reply:
@jpivarski thanks for the detail here and I think the way you've laid out where `events.Muon.compute()` would get the whole muon object and `events.Muon[events.Muon.pt > 10].eta` would only get `Muon_pt` and `Muon_eta` in the underlying nanoaod file is largely expected behavior among our users (as I understand from seeing usage patterns). It (largely) preserves the lazy interface to reading data that was there before using uproot4/awkward1.

A problem arises when we start thinking about users using histograms and tracing programs.
```python3
def processor_proxy(dx):
    h = hist.Hist(hist.axis.Regular(bins=10, start=0, stop=1, name="x"))
    h.fill(x = dx.compute())
    return h

x = ak.Array([[1.0, 2.0], [], [3.0, 4.0, 5.0]]) 
dx = dak.from_awkward(x)
# these two lines get replaced with uproot.dask in something realistic, but you get the picture...
h = processor(proxy(dx))
```

Boiled down to its essentials, this will be the most common initial pattern that users encounter with coffea 0.8 and dask awkward. Most users will have some histograms interspersed with data access, and I'm not sure from the discussion above what will happen in "tracing mode" (when, say, we are only trying to figure out what columns are desired rather than doing any real processing) when a concrete array is required and no concrete data is actually rendered (like what we would do as a preprocessing step for using spark, since you must define the columns you want to read before running if you want to be efficient in reading).

In the fullness of time `h` becomes a `dask_histogram` and the dask task graph becomes much smaller and less branchy (as the above would need to be replicated over each chunk of a dataset as we do now, largely to preserve interface). Once nothing is materialized by the user this this becomes less of an issue, I think.

Reply:
Yeah, the `.compute()`s will have to be added carefully.

With all of the above implemented, the following would download everything, but only histogram one column:

```python
nanoaod = uproot.dask("root://source.cern")

h = hist.Hist(hist.axis.Regular(bins=10, start=0, stop=1, name="pt"))

h.fill(pt = nanoaod.compute()["Muon_pt"])
```

but the following would only download one column and histogram it:

```python
h.fill(pt = nanoaod["Muon_pt"].compute())
```

Awkward 1 VirtualArrays tried to deduce what needed to be read, sometimes incorrectly, but now that `.compute()` is an explicit step, a user can use it wrong.

The same is true if it went through a restructuring step:

```python
nanoaod = uproot.dask("root://source.cern")

nanoevents = ak.zip({
    "muons": zk.zip({
        "pt": nanoaod["Muon_pt"],
        "phi": nanoaod["Muon_phi"],
        "eta": nanoaod["Muon_eta"],
        "charge": nanoaod["Muon_charge"],
    }),
    "electrons": ...,
})

h = hist.Hist(hist.axis.Regular(bins=10, start=0, stop=1, name="pt"))

h.fill(pt = nanoevents.compute().muons.pt)
```

downloads everything but only uses `"Muon_pt"`,

```python
h.fill(pt = nanoevents.muons.compute().pt)
```

downloads all of the muon data but only uses `"Muon_pt"`, and

```python
h.fill(pt = nanoevents.muons.pt.compute())
```

downloads only the `"Muon_pt"`. Scripts that upgrade to Coffea 0.8 will need to have "`.compute()`" added in various places anyway, but it will be important to add them in the right places!

Maybe dask-histogram would further simplify this:

```python
h = dask_hist.Hist(hist.axis.Regular(bins=10, start=0, stop=1, name="pt"))

h.fill(pt = nanoevents.muons.pt)

h.compute()
```

The histogram-filling would be part of the Dask graph; following that back, only the `nanoevents.muons.pt` is involved in histogram-filling, and the other buffers are untouched, so this would do the right thing with the advantage that the "`.compute()`" is only added to the final histogram, which may be easier to reason about.

Reply:
Hi Jim, that's not quite the concern!

It is very easy to tell users to put any `compute()` they may want at the end of a data-access statement as a matter of best practice, or to make a migration script (but this has a lot of possible misinterpretations of user code, so maybe it is best to tell people rather than to generate code that functions strangely compared to user's intent).

Mostly I would like the following hypothetical code to complete without error and tell me what data was touched:
```python3
proc = SomeProcessor(args)

events = NanoEventsFactory.from_root("/some/file.root", schemaclass=SomeSchema) 
# the above will use uproot.dask or dak.from_awkward, or dak.from_parquet
tracer = dak.dak.typetracer_from_form(events._form)
proc.process(tracer)
necessary_columns  = tracer.touched_data()
```

Regardless of what histograms are filled within that Processor's process.

Reply:
For debugging what was touched? You could use the same `ak._typetracer.typetracer_with_report` function that dask-awkward will be using. We'll be making that interface public at some point, but first we want to find out how much of it is going to be needed, so that we can make the public API as small as possible.

After the dust settled, the final way to use it is:

```python
result = proc.process(events)  # Coffea, ending with a dak.Array

labeled_form = copy.deepcopy(result._meta.layout.form)

labeled_form.content.form_key = ...   # label each of the nodes

new_typetracer, report = ak._typetracer.typetracer_with_report(labeled_form)

final = result._graph.run_through_graph_with(ak.Array(new_typetracer))

final._meta.layout._touch_data(recursive=True)  # important: all outputs should be labeled

print(report.data_touched)
```

  * You have to label the nodes in `labeled_form` because the buffers are the values of a dict and don't know their keys. But here, you get to specify what's important to you.
  * I don't actually know the Dask command to `run_through_graph_with` a new type-tracer, but I know that there's a mechanism for this because it's what's being done right now.
  * The step of recursively touching all the data in the final computed array is important.

But then the `report.data_touched` list contains the "`necessary_columns`" that you're asking about, a subset of the labels you assigned in the `labeled_form` step.

Is that what you're asking?

Reply:
Not only debugging what was touched, in particular for those using spark we have to pre-determine the arrays to load since spark requires a static schema when talking to root. Therefore we will need to trace the Processor once per dataset to determine the dataset schema to-be read to define the spark schema when reading data from uproot.

Here's an example, fully exploded spark workflow using spark3 and coffea 0.8 PR where you can why I need to define the dataset's schema ahead of time. It's nice since it uses mapInArrow everywhere but that does not rid us of the schema issue. I'm happy to field ideas for removing it though!:
```python3
import pyspark
from pyarrow.util import guid
import os
import os.path as osp
from coffea.processor.spark.detail import (
    _spark_initialize,
    _spark_stop,
)

spark_config = (
    pyspark.sql.SparkSession.builder.appName("spark-executor-test-%s" % guid())
    .master("local[*]")
    .config("spark.sql.execution.arrow.enabled", "true")
    .config("spark.driver.host", "127.0.0.1")
    .config("spark.driver.bindAddress", "127.0.0.1")
    .config("spark.executor.x509proxyname", "x509_u12409")
    .config("spark.sql.execution.arrow.maxRecordsPerBatch", 200000)
)

spark = _spark_initialize(
    config=spark_config, log_level="ERROR", spark_progress=False
)

filelist = {
    "ZJets": {
        "files": ["file:" + osp.join(os.getcwd(), "tests/samples/nano_dy.root")],
        "treename": "Events",
    },
    "Data": {
        "files": [
            "file:" + osp.join(os.getcwd(), "tests/samples/nano_dimuon.root")
        ],
        "treename": "Events",
    },
}

columns = ["dataset", "files", "treename"]
reshaped = []
for key, item in filelist.items():
    reshaped.append((key, *(val for val in item.values())))

df = spark.createDataFrame(reshaped, columns)

def explode_dataset(iterator):
    import pyarrow as pa
    import awkward as ak
    for batch in iterator:
        array = ak.from_arrow(batch)
        expander = ak.zeros_like(array.files[0], dtype=int)
        array = ak.zip({"dataset": array.dataset[expander], "file": array.files[0], "treename": array.treename[expander]}, depth_limit=1)
        fields = ak.fields(array)
        yield pa.RecordBatch.from_arrays([array[field] for field in fields], fields)

from pyspark.sql.types import StructType, StructField, StringType, BinaryType, IntegerType, LongType, FloatType, ArrayType

outSchema = StructType([       
    StructField('dataset', StringType(), True),
    StructField('file', StringType(), True),
    StructField('treename', StringType(), True)
])

df = df.mapInArrow(explode_dataset, outSchema)

columns = ["Muon_pt", "Muon_eta", "Muon_phi", "Muon_mass", "Muon_charge", "nMuon"]

def datasetdef_to_columns(iterator):    
    import pyarrow as pa
    import awkward as ak
    import numpy as np
    import uproot
    for batch in iterator:
        dsinfo = ak.from_arrow(batch)
        for i in range(len(batch)):
            with uproot.open({dsinfo.file[i]: dsinfo.treename[i]}) as t_in:
                ak_arrays = t_in.arrays(columns)
                ak_arrays["nMuon"] = ak.values_astype(ak_arrays["nMuon"], np.int64)
                expander = ak.zeros_like(len(ak_arrays)*[0], dtype=int)
                ak_arrays["dataset"] = dsinfo.dataset[expander]
                pa_arrays = ak.to_arrow_table(ak_arrays, list_to32=True, string_to32=True, extensionarray=False).to_batches()
                for out_batch in pa_arrays:
                    yield out_batch

columnTypes = []
for column in columns:
    if column == "nMuon":
        columnTypes.append(StructField(column, LongType(), False))
    elif column == "Muon_charge":
        columnTypes.append(StructField(column, ArrayType(IntegerType()), False))
    else:
        columnTypes.append(StructField(column, ArrayType(FloatType()), False))
    
columnsSchema = StructType(columnTypes + [StructField('dataset', StringType(), False)])

df = df.mapInArrow(datasetdef_to_columns, columnsSchema)

from coffea.nanoevents import NanoEventsFactory, NanoAODSchema
from coffea.processor.test_items import NanoEventsProcessor
from coffea.nanoevents.mapping import SimplePreloadedColumnSource

import lz4.frame
import pickle

nano_schema = NanoAODSchema
lz4_clevel = 1
processor_instance = NanoEventsProcessor()

def columns_to_histograms(iterator):
    global processor_instance, lz4_clevel, nano_schema

    import awkward as ak
    import pyarrow as pa

    for batch in iterator:
        arrays = ak.from_arrow(batch)

        fields_no_meta = ak.fields(arrays)
        fields_no_meta.remove("dataset")
        arrays_no_meta_no_none = {field: ak.fill_none(arrays[field],0) for field in fields_no_meta}
        
        src = SimplePreloadedColumnSource(arrays_no_meta_no_none, None, len(arrays), object_path='/Events')

        events = NanoEventsFactory \
            .from_preloaded(src, metadata={'dataset': arrays.dataset[0]}, schemaclass=nano_schema) \
            .events()  

        vals = processor_instance.process(events)

        valsblob = lz4.frame.compress(pickle.dumps(vals), compression_level=lz4_clevel)
        
        out_batches = ak.to_arrow_table(valsblob, bytestring_to32=True, extensionarray=False).to_batches()        
        for out_batch in out_batches:
            yield out_batch

histoSchema = StructType([
    StructField('histos',BinaryType(),False)
])

df = df.mapInArrow(columns_to_histograms, histoSchema)

df.show()
```

Reply:
Okay, so you do need to use this outside of dask-awkward: there will be two direct users of it, dask-awkward and Coffea. That will influence the design of the public API, after we see how this works in practice.

Assuming that you have a Processor function named `process`, which takes an input `nanoevents` with known Form `nanoevents_form` and produces results through histogram-filling side-effects, this should do it:

```python
labeled_form = copy.deepcopy(nanoevents_form)

labeled_form.content.form_key = ...   # label each of the nodes

nanoevents_tt, report = ak._typetracer.typetracer_with_report(labeled_form)

process(nanoevents_tt)

print(report.data_touched)
```

Unlike the Dask case, there's no final `materialized` array that users might do subsequent operations on and be surprised that it lacks or has dummy values for some field that they asked for but didn't touch in the asking. So there's no need to call `_touch_data(recursive=True)` on a "final result."

If the histogram functions try to get data from the `nanoevents_tt` by converting the arrays into NumPy arrays (e.g. what hist does), the `TypeTracerArray` would complain with

```
AssertionError: bug in Awkward Array: attempt to convert TypeTracerArray into a concrete array
```

because we were assuming that only we would be using `TypeTracerArrays` directly; dask-awkward doesn't try to coerce it into a NumPy array, it just puts the `ak.to_numpy` operation in the graph.

If you have control over the histogram `fill` function (Coffea histograms, wrapped boost-histogram/hist, or some Spark histogram), you could replace the

```python
ak.to_numpy(awkward_data_to_histogram)  # or
np.asarray(awkward_data_to_histogram)
```

with

```python
awkward_data_to_histogram.layout._touch_data(recursive=True)
```

Otherwise, we'll need to make `TypeTracerArray.__array__` not raise an error, call `_touch_data` on its own, and return a dummy array. Not having this assertion removes one of the guardrails in our own development, but most of that development is done now, so that would be okay.